tools/codegen/pkg/modelgen/model.go

```diff
@@ -0,0 +1,265 @@
+	}
+}
+
+// Build field information about the Constructor message.
+func (m *Model) addInstanceFieldFromConstructor(parser *FileDescriptorSetParser, cnstrDesc *Descriptor) {
+	m.ConstructorFields = make([]fieldInfo, 0)
+	for _, fieldDesc := range cnstrDesc.Field {
+		fieldName := CamelCase(fieldDesc.GetName())
+		// Name field is a reserved field that will be injected in the Instance object. The user defined
+		// Constructor should not have a Name field, else there will be a name clash.
+		// 'Name' within the Instance object would represent the name of the Constructor:instance_name
+		// specified in the operator Yaml file.
+		if fieldName == "Name" {
+			// TODO add line number for the field.
+			m.addError(cnstrDesc.file.GetName(), unknownFile, "Constructor message must not contain the reserved filed name '%s'", fieldDesc.GetName())
+			continue
+		}
+		typename := parser.GoType(cnstrDesc.DescriptorProto, fieldDesc)
+		// TODO : Can there be more than one expressions in a type for a field in Constructor ?
+		typename = strings.Replace(typename, fullNameOfExprMessageWithPtr, "interface{}", 1)
+		m.ConstructorFields = append(m.ConstructorFields, fieldInfo{Name: fieldName, Type: typeInfo{Name: typename}})
+	}
+}
+
+func (m *Model) addImports(parser *FileDescriptorSetParser, fileDescriptor *FileDescriptor, typDesc *Descriptor) {
+	usedPackages := getReferencedPackagesWithinType(parser, typDesc)
+	m.Imports = make([]string, 0)
+	for _, s := range fileDescriptor.Dependency {
+		fd := parser.fileByName(s)
+		// Do not import our own package.
+		if fd.PackageName() == parser.packageName {
+			continue
+		}
+		filename := fd.goFileName()
+		// By default, import path is the dirname of the Go filename.
+		importPath := path.Dir(filename)
+		if substitution, ok := parser.ImportMap[s]; ok {
+			importPath = substitution
+		}
+
+		pname := goPackageName(fd.GetPackage())
+		if !contains(usedPackages, pname) {
+			pname = "_"
+		}
+		m.Imports = append(m.Imports, pname+" "+strconv.Quote(importPath))
+	}
+}
+
+func getRequiredMsg(fdp *FileDescriptor, msgName string) (*Descriptor, bool) {
+	var cstrDesc *Descriptor = nil
+	for _, desc := range fdp.desc {
+		if desc.GetName() == msgName {
+			cstrDesc = desc
+			break
+		}
+	}
+
+	return cstrDesc, cstrDesc != nil
+}
+
+func getReferencedPackagesWithinType(parser *FileDescriptorSetParser, typDesc *Descriptor) []string {
+	result := make([]string, 0)
+	for _, fieldDesc := range typDesc.GetField() {
+		getReferencedPackagesByField(parser, fieldDesc, typDesc.PackageName(), &result)
+	}
+
+	return result
+}
+
+func getReferencedPackagesByField(parser *FileDescriptorSetParser, fieldDesc *descriptor.FieldDescriptorProto, parentPackage string, referencedPkgs *[]string) {
+	if *fieldDesc.Type != descriptor.FieldDescriptorProto_TYPE_MESSAGE && *fieldDesc.Type != descriptor.FieldDescriptorProto_TYPE_ENUM {
+		return
+	}
+	refDesc := parser.ObjectNamed(fieldDesc.GetTypeName())
+	if d, ok := refDesc.(*Descriptor); ok {
```

Comment:
not a blocker, just curious: is there a way to combine/collapse some of this logic here? it feels like there should a better way to structure this... something along the lines of:

```
desc, ok := refDesc.(*Descriptor)
if !ok || desc.GetOptions().GetMapEntry() {
      pname := ...
      return
}
if fmt.Sprintf(...) ... {
      return
}
if desc.GetOptions().GetMapEntry() {
      ...
}
```